pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -605,7 +601,7 @@ func addEndpointSubset(subsets []v1.EndpointSubset, pod *v1.Pod, epa v1.Endpoint
 			Ports:     ports,
 		})
 		readyEps++
-	} else if shouldPodBeInEndpoints(pod) {
+	} else { // unready pod
```

Comment:
This doesn't seem quite right. With EndpointSlices we don't include any endpoints unless the `ShouldPodBeInEndpointSlice` function returns true:

https://github.com/kubernetes/kubernetes/blob/114cdea709b7046c439d9c4088cb9ab75105a0f0/pkg/controller/endpointslice/reconciler.go#L157-L159

If that function returns true, we then separately determine if the endpoint should be marked as ready:

https://github.com/kubernetes/kubernetes/blob/da23196df6d519c657a27c68932c62aa35e87dc5/pkg/controller/endpointslice/utils.go#L45-L49

I think we'd want to largely copy that logic here, but instead of determining the state of the ready bool, this logic would determine if the endpoint would be added to `addresses` or `notReadyAddresses` in the Endpoints object.

I'm still trying to think through any unwanted interactions with the terminating endpoints condition. As far as I can tell, we should be safe because terminating pods are already handled in the Endpoints controller here:

https://github.com/kubernetes/kubernetes/blob/da23196df6d519c657a27c68932c62aa35e87dc5/pkg/controller/endpoint/endpoints_controller.go#L423-L425